File: c/binary_search.c
#include "algo.h"
#include <stddef.h>
#include <stdio.h>

BS_IntPair binary_search(int *array, int array_length, int item) {
  int iterations = 0;

  if (array == NULL) {
    return (BS_IntPair){.iter = 0, .pos = -1};
  }

  int low = 0;
  int high = array_length - 1;
  while (low < high) {
    iterations += 1;

    int mid = (low + high) / 2;
    printf("mid: %d\n", mid);
    int item_at_mid = array[mid];
    if (item_at_mid > item) {
      high = mid - 1;
    } else if (item_at_mid < item) {
      low = mid + 1;
    } else {
      return (BS_IntPair){.iter = iterations, .pos = mid};
    }
  }

  if (array[low] == item) {
    return (BS_IntPair){.iter = iterations, .pos = low};
  }

  return (BS_IntPair){.iter = iterations, .pos = -1};
}

void print_BS_IntPair(BS_IntPair *res) {
  if (res == NULL) {
    return;
  }

  printf("BS_IntPair { iter: %d, pos: %d }\n", res->iter, res->pos);
}

File: ts/src/binarySearch.ts

type Index = number;
type NumOfIterations = number;

function binarySearch(list: number[], item: number): [Index | null, NumOfIterations]{
    let numOfIterations = 0;

    if (list.length == 0){
        return [null, numOfIterations];
    }

    let low = 0;
    let high = list.length - 1;

    while (low < high) {
        numOfIterations += 1;

        const mid = Math.floor((low + high) / 2);
        const itemAtMid = list[mid];
        console.log(`mid item: ${itemAtMid}`)
        if (itemAtMid == undefined){
            return [null, numOfIterations];
            // will actually never be null since 'high' will always be valid
        }

        if (itemAtMid > item) {
            high = mid - 1;
        } else if (itemAtMid < item) {
            low = mid + 1;
        } else {
            return [mid, numOfIterations];
        }
    }

    if (list[low] == item){
        return [low, numOfIterations];
    }else {
        return [null, numOfIterations];
    }
}

const inputList = [1, 2, 3, 4, 5, 6, 8, 10, 11, 12, 13];
const item = 13;
const [res, iters] = binarySearch(inputList, item);
console.log(`index of ${item}: ${res} (with ${iters} iters)`);
